archunit/src/jdk9test/java/com/tngtech/archunit/core/importer/ClassFileImporterLambdaAccessesTest.java

```diff
@@ -0,0 +1,463 @@
+package com.tngtech.archunit.core.importer;
+
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import com.tngtech.archunit.core.domain.AccessTarget.MethodCallTarget;
+import com.tngtech.archunit.core.domain.JavaAccess;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.JavaConstructorCall;
+import com.tngtech.archunit.core.domain.JavaConstructorReference;
+import com.tngtech.archunit.core.domain.JavaFieldAccess;
+import com.tngtech.archunit.core.domain.JavaMethod;
+import com.tngtech.archunit.core.domain.JavaMethodCall;
+import com.tngtech.archunit.core.domain.JavaMethodReference;
+import org.assertj.core.api.Condition;
+import org.junit.Test;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.tngtech.archunit.core.domain.Formatters.formatNamesOf;
+import static com.tngtech.archunit.core.domain.JavaConstructor.CONSTRUCTOR_NAME;
+import static com.tngtech.archunit.core.domain.properties.HasName.Utils.namesOf;
+import static com.tngtech.archunit.core.importer.JavaClassDescriptorImporterTestUtils.isLambdaMethodName;
+import static com.tngtech.archunit.testutil.Assertions.AccessesAssertion.access;
+import static com.tngtech.archunit.testutil.Assertions.assertThat;
+import static com.tngtech.archunit.testutil.Assertions.assertThatAccess;
+import static com.tngtech.archunit.testutil.Assertions.assertThatAccesses;
+import static com.tngtech.archunit.testutil.Assertions.assertThatCall;
+import static java.util.stream.Collectors.toSet;
+
+@SuppressWarnings("ResultOfMethodCallIgnored")
+public class ClassFileImporterLambdaAccessesTest {
+    @Test
+    public void imports_method_call_from_lambda_without_parameter() {
+        class Target {
+            void target() {
+            }
+        }
+
+        @SuppressWarnings("unused")
+        class Caller {
+            private Target target;
+
+            Runnable call() {
+                return () -> target.target();
+            }
+        }
+
+        JavaClasses classes = new ClassFileImporter().importClasses(Target.class, Caller.class);
+        JavaMethodCall call = getOnlyElement(classes.get(Caller.class).getMethodCallsFromSelf());
+
+        assertThatCall(call).isFrom("call").isTo(Target.class, "target");
```

Comment:
While I agree that there should be a dependency from `call` to `Target#target()`, I'm not sure whether this should be a `JavaMethodCall` with _replaced origin_ (and furthermore one that cannot be distinguished from an actual method call).

If it was only for the dependency, this could also be established via some `JavaCodeUnit#getLambdas()`.

If you really want to provide the lambdas via `getMethodCallsFromSelf()`, I'd prefer to have some kind of marker to be able to recognize (and possibly filter) rewritten method calls, e.g. via a subclass `JavaMethodCallForLambda extends JavaMethodCall` or a property `hasRewrittenOriginForLambda` on `JavaMethodCall`.

What strikes me most about the proposed solution is that 
```
return () -> target.target();
```
behaves differently from
```
return new Runnable() { 
    public void run() {
        target.target();
    }
};
```
– but I don't see much difference from an architectural point of view.

Reply:
As discussed I added the additional info `JavaAccess.isDeclaredInLambda()` which now provides this context information. The problem with a separate subclass would have been, that we don't just have method calls from lambdas, we have any access that could now be from within a lambda. So we would have to "blow up" the whole domain hierarchy if we would really want to model this in this way. I think for most common cases it is okay for users to see accesses declared within lambdas the same way as accesses declared outside. It might be slightly inconsistent with anonymous classes, but in the end within a lambda there just is no other class involved, while for an anonymous class there is :man_shrugging: That's more the fault of the JLS :wink: